fix: Fixes Issue #144. Extrinsics with high balance restarts webpage

ltfschoen opened the pull request:
**Screenshots**:

* Default value of 1 testnet DOT
<img width="341" alt="screen shot 2018-08-14 at 5 48 03 pm" src="https://user-images.githubusercontent.com/6226175/44102782-cb8a4950-9fea-11e8-9be4-852f45d16585.png">

* Input validation error shown when no input value
<img width="368" alt="screen shot 2018-08-14 at 5 48 14 pm" src="https://user-images.githubusercontent.com/6226175/44102787-cf8b2ab0-9fea-11e8-90d6-d5fc898f23d9.png">

* Input validation error shown when user inputs non-numeric values
<img width="360" alt="screen shot 2018-08-14 at 5 48 27 pm" src="https://user-images.githubusercontent.com/6226175/44102821-e61fea72-9fea-11e8-92ab-e6854a903dbe.png">

* Throw exception input value into the `isValidBalance` function is in scientific notation
<img width="709" alt="screen shot 2018-08-14 at 5 49 06 pm" src="https://user-images.githubusercontent.com/6226175/44102848-f5bf5c10-9fea-11e8-9470-ecbce0bf259a.png">

* Maximum amount of digits allowed for the balance input value
<img width="532" alt="screen shot 2018-08-14 at 5 51 28 pm" src="https://user-images.githubusercontent.com/6226175/44102940-344d5842-9feb-11e8-95ed-4b8d0587b3c6.png">

**References:**
* Previous PR that was closed https://github.com/polkadot-js/apps/pull/168
* Original PR from my fork that was closed in favour of a branch from the upstream repo https://github.com/polkadot-js/apps/pull/146

ltfschoen commented:
@jacogr I've finished incorporating your review comments. If there's an error in the user input we shown them an error message notification. If there's no error and they enter a value in either scientific notation (i.e. 3.4e3) or exponential notation (i.e. 3.4e+3) then I we shown them an information notification below the input field with the converted equivalent integer value (i.e. 3400). Note that I allow decimal places only when they enter a value in scientific notation or exponential notation. If they enter an integer value (i.e. 3400) it shows an information notification with the converted equivalent value in scientific notation (i.e. 3.4e3) below. So the PR supports E-Notation #224.

The changes also require changes to polkadot-js/api, and I have created an PR here [https://github.com/polkadot-js/api/pull/139](https://github.com/polkadot-js/api/pull/139)

ltfschoen commented:
@jacogr I've tried implementing InputNumber again, hopefully this is what you meant?

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,59 @@
+          type='text'
+          withLabel={withLabel}
+        />
+        <Notification
```
Notification should be rendered on `Input` itself, not really on the wrapping component. (It is useful for all inputs, e.g. Password, Balance, normal, etc.)

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    return (
+      <div className='ui--Notifications'>
```
This one we would want to add the `className` & `style` as passed-in as well. (Sinmce it is a base component and we may want to apply overrides on a per-use basis)

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,41 @@
+          info={!error ? !!info : undefined}
+        >
+          {error}
+          {!error ? info : null}
```
Lots of `!error` here. Rather make a variable `isError` and use that, i.e.

```
const isError = !!error;
const isInfo = !isError && !!info;
```

(The logic here is not easy to follow, some mental gymnastics required.)

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+import withApi from '@polkadot/ui-react-rx/with/api';
+
+import isValidBalance from '../../util/isValidBalance';
+import Bare from './Bare';
```
Move this import to be last, i.e. first imports external to us, then our import.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+    const { error, info } = this.state;
+    const defaultValue = new BN((value as BN).toString(10) || '0').toString(10);
+
+    const maxLengthForLatestChainSpec = 39;
```
This is going to create issues. :( For InputNumber, pass through the bitLength and calculate the number there. i.e. 

```
const maxValue = new BN(2).pow(bitLength).subn(1);
const maxLength = maxValue.toString().length;`
```

Additionally, the input value can now actually be checked against this maxValue to see that if it is still indeed in-range.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+          onKeyDown={this.onKeyDown}
+          onKeyUp={this.onKeyUp}
+          placeholder={t('account.balance.placeholder', {
+            defaultValue: 'Between 1 testnet DOT and the available testnet DOT balance (minus 1) of the account'
```
We should no mention testnet, keep it generic otherwise we need to come back to this. Additionally, we can allow 0 as an input. (Just not negative). Why "minus 1"

ltfschoen replied:
Ok. I can't recall the reason for the "minus 1". It may have been because of errors in the Polkadot Node logs when I tried sending the full account balance, but I'll have to check again once I revive Alice.
Why would we allow them to enter 0 as an input?

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+    value = value.split(' ').join('');
+
+    try {
+      const { isValid, errorMessage, infoMessage, inputConvertedFromScientificNotation } = isValidBalance(value);
```
`inputConvertedFromScientificNotation` this is getting very long... something like `isScientific` would do just as well.

jacogr replied:
I'm probably even misunderstanding the use here - so make it clear. If it is a value, call it `value`, if it is a bool, call it `is<Something>`

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+        info: isValid && infoMessage ? t(infoMessage) : ''
+      });
+
+      if (!onChange) return;
```
No single-line returns, add `{}`

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+      if (!onChange) return;
+
+      let newValue: BN;
+      if (inputConvertedFromScientificNotation) {
```
Blank line before.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+        value: newValue
+      });
+    } catch (e) {
+      console.error('error: ', e);
```
Just console.error(error) would do (Name it `error`, no need to shorten)

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+    const zeroKey = 48;
+    const zeroNumpadKey = 96;
+
+    const regexE = /[e]/gi;
```
Remove extra blank line.

All these can be constants that are defined on top of the file (and named accordingly, i.e. `KEY_E` or `KEY_PLUS`). As it stands each time the method fires the variables are created in-scope.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+    const inputCursorIndex = event.target.value.slice(0, event.target.selectionStart).length;
+
+    // allow these keys
+    if ([escapeKey, enterKey, backspaceKey, delKey, tabKey, ctrlKey, altKey, cmdKey, eKey, arrowLeftKey,
```
`KEYS_ALLOWED` (Once again, we don't need to re-create the array each time, it is file constant)

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+      ((event.ctrlKey || cmdKey) && event.which === aKey) || // select all
+      ((event.ctrlKey || cmdKey) && event.which === cKey) || // copy
+      ((event.ctrlKey || cmdKey) && event.which === vKey) || // paste
+      ((event.ctrlKey || cmdKey) && event.which === xKey) || // cut
```
`(event.ctrlKey || cmdKey)` is re-used, so something like this would work as well -

``(event.ctrlKey || cmdKey) && [aKey, cKey, vKey, xKey].includes(vKey)`

(Once again in the above example, just make sure we have the proper constant arrays defined.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+
+  onKeyUp = (event: any): void => {
+    // if user inputs the value of 'E', replace it with lowercase 'e'
+    const regexE = /[E]/gi;
```
file-based constant.

ltfschoen replied:
Sorry, I'm not sure what you mean. Would you prefer I change it to say `event.target.value.indexOf('E')`?

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,97 @@
+  const matchEPlusBeforeNumber = input.match(/(e\+)(?=[0-9])/ig);
+
+  if (matchE && matchE.length > 1 || matchPlus && matchPlus.length > 1) {
+    return { isValid: false, errorMessage: 'Scientific notation may only contain one instance of \'e\' for scientific notation or exponential with \'e+\'' };
```
- Each prop of the object on a seperate line.
- All messages need to be translatable, these are not

jacogr replied:
(The above `t(errorMessage)` approach is not correct, it will just create havoc for a translation pass.)

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,97 @@
+      return {
+        isValid: true,
+        infoMessage: inputConvertedFromScientificNotation,
+        inputConvertedFromScientificNotation: inputConvertedFromScientificNotation
```
Simpler to do -

```
return {
        isValid: true,
        infoMessage,
        inputConvertedFromScientificNotation
}
```

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,97 @@
+    if (matchEPlus && !inputConvertedFromScientificNotationBN.lt(maxBN128Bit)) {
+      return { isValid: false, errorMessage: 'Balance value after converting from exponential notation exceeds maximum for 128 bit' };
+    } else if (!inputConvertedFromScientificNotationBN.lt(maxBN128Bit)) {
+      return { isValid: false, errorMessage: 'Balance value after converting from scientific notation exceeds maximum for 128 bit' };
```
Same comments are earlier in the file - between this and all the preceding lines the same issue occurs. (Formatting & non-translatable messages)

ltfschoen replied:
I've refactored the code so it returns a translated `errorMessage`, but I'm also now returning with it an untranslated `errorMessageUntranslated` so it doesn't interfere with the unit tests as I had difficulty trying to mock the translate function

jacogr commented on packages/ui-app/src/util/isValidBalanceExpectedResponses.ts:
```diff
@@ -0,0 +1,32 @@
+import { IsValidWithMessage } from './types';
+
+export enum ErrorMessage {
+  BalanceMustContainOnlySingleEOrPlus = 'Scientific notation may only contain one instance of \'e\' for scientific notation or exponential with \'e+\'',
```
No. All messages needs to be translatable when they are used on the UI. As it stand, these are not.

ltfschoen replied:
I'm only using `ErrorMessage` as a helper for unit testing purposes in isValidBalance.spec.js,  but not on the UI

jacogr commented on packages/ui-app/src/util/isValidBalanceExpectedResponses.ts:
```diff
@@ -0,0 +1,32 @@
+  BalanceMinimumRequired = 'Balance to transfer in DOTs must be greater than or equal to one',
+  BalanceMustBeFinite = 'Balance to transfer in DOTs must not be infinite',
+  BalanceWithDecimalsOnlyWithE = 'Decimals points are only allowed in scientific notation by using an \'e\' (i.e. 3.4e38) or exponential with \'e+\'',
+  BalanceExceedsMaximum = 'Balance exceeds maximum for 128 bit',
```
No. Needs to be specific to chain and specific to the number required. (Translatable param strings cater for this)

jacogr commented on packages/ui-app/src/util/isValidBalanceExpectedResponses.ts:
```diff
@@ -0,0 +1,32 @@
+  BalanceWithDecimalsOnlyWithE = 'Decimals points are only allowed in scientific notation by using an \'e\' (i.e. 3.4e38) or exponential with \'e+\'',
+  BalanceExceedsMaximum = 'Balance exceeds maximum for 128 bit',
+  BalanceExceedsMaximumScientificNotation = 'Balance value after converting from scientific notation exceeds maximum for 128 bit',
+  BalanceExceedsMaximumExponentialNotation = 'Balance value after converting from exponential notation exceeds maximum for 128 bit'
```
There are all overly long and Java-ish: They don't all need the `Balance` prefix, duplication end extra words can be omitted, e.g. `Exceeds` add no real value.

jacogr commented on packages/ui-app/src/util/isValidBalanceExpectedResponses.ts:
```diff
@@ -0,0 +1,32 @@
+    response.infoMessage = infoMessage;
+  }
+  if (inputConvertedFromScientificNotation) {
+    response.inputConvertedFromScientificNotation = inputConvertedFromScientificNotation;
```
As above, rename this overly long var

jacogr commented on packages/ui-app/src/util/scientificNotationToNumber.ts:
```diff
@@ -0,0 +1,44 @@
+      let l = Math.abs(Number(e)); // get the number of zeros
+      if (isNaN(l)) {
+        return {
+          errorMessage: 'Exponential part of scientific notation is invalid',
```
Non-translatable.

jacogr commented on packages/ui-app/src/util/scientificNotationToNumber.ts:
```diff
@@ -0,0 +1,44 @@
+    const zero = '0';
+    const parts = String(num).toLowerCase().split('e'); // split into coeff and exponent
+    let e = parts.pop(); // store the exponential part
+    if (e) {
```
Blank line before.

jacogr commented on packages/ui-app/src/util/scientificNotationToNumber.ts:
```diff
@@ -0,0 +1,44 @@
+    } else {
+      return {
+        errorMessage: 'Exponential part of scientific notation is missing',
+        inputConvertedFromScientificNotation: undefined
```
Overly long field, non-translatable string.

jacogr commented on packages/ui-app/src/util/scientificNotationToNumber.ts:
```diff
@@ -0,0 +1,44 @@
+    }
+  }
+  return {
+    errorMessage: undefined,
```
No need to add the field here if not used. Just leave out instead of saying `undefined`

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,192 @@
+      const { isValid, errorMessage, infoMessage, inputConvertedFromScientificNotation } = isValidBalance(value);
+
+      this.setState({
+        error: !isValid && errorMessage ? t(errorMessage) : '',
```
???

The way this is done -

`t(<name-for-the-string>, { defaultValue: <default-english-value> })`

That is the only way it will make sense to translate since `name-for-the-string` gives content, where/what and `default-english-value` shows the translator what the original was.

jacogr replied:
EDIT: Technically the above would work, however where the concern comes in - the mapping now matches against the full text, instead of a static key that should not change. If we now make any changes to that string (which is used in translation matching), the already translated text breaks. (So something as simple as a type breaks anything that has been translated for that text.)

Always stick to key + defaultValue everywhere.

ltfschoen commented:
@jacogr I've addressed the review comments except for the following:

1) We should no mention testnet, keep it generic otherwise we need to come back to this. Additionally, we can allow 0 as an input. (Just not negative). Why "minus 1"

  * **RESPONSE** - Ok. I can't recall the reason for the "minus 1". It may have been because of errors in the Polkadot Node logs when I tried sending the full account balance, but I'll have to check again once I revive Alice.
	Why would we allow them to enter 0 as an input?
	Sorry, rephrase... why would we allow them to enter just a value of 0 as an input (i.e. submit a transaction with a Balance value of 0)?

2) inputConvertedFromScientificNotation this is getting very long... something like isScientific would do just as well.
I'm probably even misunderstanding the use here - so make it clear. If it is a value, call it value, if it is a bool, call it is<Something>

  * **RESPONSE** - Renamed to `num`

3) (event.ctrlKey || cmdKey) is re-used, so something like this would work as well -
``(event.ctrlKey || cmdKey) && [aKey, cKey, vKey, xKey].includes(vKey)`
(Once again in the above example, just make sure we have the proper constant arrays defined.

  * **RESPONSE** - I've encountered a bug where it allows user to enter repeatedly the following non-numeric keys still: a, c, v, x multiple times (since I can't seem to only allow those characters to be presseed when in combination with them pressing CMD or CTRL). Any ideas how I can resolve?

4) file-based constant.

  * **RESPONSE** - Sorry, I'm not sure what you mean. Would you prefer I change it to say `event.target.value.indexOf('E')`?

5) Same comments are earlier in the file - between this and all the preceding lines the same issue occurs. (Formatting & non-translatable messages)

  * **RESPONSE** - I've refactored the code so it returns a translated errorMessage, but I'm also now returning with it an untranslated errorMessageUntranslated along with it so it doesn't interfere with the unit tests. I tried to avoid doing this but had difficulty trying to mock the translate function

6) No. All messages needs to be translatable when they are used on the UI. As it stand, these are not.

  * **RESPONSE** - I'm only using `ErrorMessage` as a helper for unit testing purposes in isValidBalance.spec.js,  but not on the UI

jacogr commented:
1. You are thinking of Balance as purely used in transfers - it is not. It is used when making proposals, starting referendums and elsewhere as a value that gets added to the `reservedBalance` property of the account. So any value (valid for decoding) is possible.

2. Ok.

3. Debugging it is the best option here. Without doing it myself, cannot provide pointers.

4. file based constants, i.e instead of doing

```
function something () {
  // only used for queries, never modified
  const constant = [...];
  // test on constant
}
```

which now gets created as a new variable (and the garbage collected when out of scope) each time the function is called. You would do -

```
const CONSTANT = [...];

function something () {
  // test on CONSTANT
  ...
}
```

5. Ok, will take a look. (Obviously if used in tests, it would be better to have enum values as numbers instead of doing string checks)

6. As above.

ltfschoen commented:
@jacogr I've restored functionality allowing input of 0. I fixed a bug I found which occurred when a user enters (for scientific or exponential notation purposes) a decimal value that was too low such that its byteLength was over 128. I've fixed the bug so users cannot enter x,c,v,a characters but can still copy/cut/paste/select all. I've added some enzyme simulation tests.

jacogr commented on packages/app-extrinsics/src/Extrinsic.tsx:
```diff
@@ -82,7 +84,9 @@ class Extrinsic extends React.PureComponent<Props, State> {
           isValid &&
           !isUndefined(values[index]) &&
           !isUndefined(values[index].value) &&
+          (values[index].type === 'Balance' ? checkValueBitLength(values[index].value as BN) : true) &&
```
Nested condition, would make sense to make the code clean and split this condition out into `isValidLength` and then just `&& isValidLength`

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -25,18 +28,18 @@ type Props = BareProps & {
   label?: any, // node?
   max?: any,
   min?: any,
+  maxLength?: number,
```
Alphabetical ordering.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,177 @@
+          onKeyDown={this.onKeyDown}
+          onKeyUp={this.onKeyUp}
+          placeholder={t('account.balance.placeholder', {
+            defaultValue: 'Between 1 DOT and the available DOT balance of the account'
```
Can be 0. (Additionally this now does not match up with the default shown above)

jacogr commented on packages/ui-app/src/util/chainSpec.ts:
```diff
@@ -0,0 +1,26 @@
+import BN from 'bn.js';
```
License header.

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,237 @@
+  let warnMessageUntranslated: string = '';
+
+  // check value is valid number, whether it be scientific notation, exponential (which allow decimal)
+  if (isNaN(parseFloat(input))) {
```
This will fail for very large values >64bits. (JS numbers cannot handle it)

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  // always a string from <input type='number'> but leave as failsafe
+  if (!(typeof input === 'string')) {
```
Use `isString`

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  // check value is not decimal and greater than or equal to one, whether it be scientific notation, exponential (which allow decimal)
+  if (Number(parseFloat(input)) < 1 && Number(parseFloat(input)) !== 0) {
```
We are using BN for a reason - JS cannot handle these. (Until BigInt is available)

jacogr commented on packages/ui-app/src/util/isValidBalanceExpectedResponses.ts:
```diff
@@ -0,0 +1,57 @@
+import { TranslationFunction } from 'i18next';
+
+export enum ErrorMessage {
+  MustBeStringType = 'Balance input value must be valid type',
```
As soon as a file is available that 

(a) is not a test
(b) has naked strings

We end up with future issues.

jacogr commented on packages/ui-app/src/util/isValidBalanceExpectedResponses.ts:
```diff
@@ -0,0 +1,57 @@
+  if (warningMessageUntranslated) {
+    response.warnMessageUntranslated = warningMessageUntranslated;
+  }
+  if (infoMessage) {
```
Blank lines to denote blocks - for the whole function.

jacogr commented on packages/ui-app/src/util/scientificNotationToNumber.ts:
```diff
@@ -0,0 +1,63 @@
+          errorMessage: t(errorMessageKey, {
+            defaultValue: errorMessageUntranslated
+          }),
+          errorMessageUntranslated: errorMessageUntranslated
```
Why? Only always translated strings.

jacogr commented on packages/ui-app/src/util/scientificNotationToNumber.ts:
```diff
@@ -0,0 +1,63 @@
+        errorMessage: t(errorMessageKey, {
+          defaultValue: errorMessageUntranslated
+        }),
+        errorMessageUntranslated: errorMessageUntranslated
```
As above.

jacogr commented on packages/ui-app/src/util/types.d.ts:
```diff
@@ -12,3 +14,19 @@ export type DropdownOption = {
+export type IsValidWithMessage = {
+  isValid: boolean,
+  errorMessage?: TranslationFunction,
+  errorMessageUntranslated?: string,
```
No. Only always translated strings everywhere.

ltfschoen commented:
Progress update:

**Review comments addressed:**
* Refactor nested condition into `isValidLength` instead of `(values[index].type === 'Balance' ? checkValueBitLength(...`
* Fix alphabetical ordering
* Change to `Between 0 DOT` instead of `Between 1 DOT`
* Add License heaer to chainSpec.ts
* Replace `if (!(typeof input === 'string')) ` with `isString`
* Add blank lines for all blocks in isValidBalanceExpectedResponses.ts

**TODO** - still have to address the following comments mentioned above:
  * This will fail for very large values >64bits. (JS numbers cannot handle it): `if (isNaN(parseFloat(input))) {`
  * We are using BN for a reason - JS cannot handle these. (Until BigInt is available) `if (Number(parseFloat(input)) < 1 && Number(parseFloat(input)) !== 0) {`
  * As soon as a file is available that (a) is not a test (b) has naked strings. We end up with future issues. Regarding:
  `MustBeStringType = 'Balance input value must be valid type',`
  * Why? Only always translated strings in ui-app/src/util/scientificNotationToNumber.ts: `errorMessageUntranslated: errorMessageUntranslated`
  * No. Only always translated strings everywhere. Re ui-app/src/util/types.d.ts: `errorMessageUntranslated?: string,`

ltfschoen commented:
@jacogr I've made simplifying changes to just focus on the issue at hand. I've removed from this PR any support for input of scientific notation (we can address that in a separate issue and leverage lessons learnt from this task)

jacogr commented on packages/app-extrinsics/src/Extrinsic.tsx:
```diff
@@ -73,6 +75,10 @@ class Extrinsic extends React.PureComponent<Props, State> {
 
   nextState (newState: State): void {
+    const isValidLength = (values: Array<RawParam>, index: number) => {
+      return (values[index].type === 'Balance' ? checkValueBitLength(values[index].value as BN) : true);
```
This would read a bit better split,

```
const isValidLength = (values: Array<RawParam>, index: number) =>
      return values[index].type === 'Balance'
          ? checkValueBitLength(values[index].value as BN)
           : true;
```

jacogr commented on packages/app-extrinsics/src/Extrinsic.tsx:
```diff
@@ -73,6 +75,10 @@ class Extrinsic extends React.PureComponent<Props, State> {
   nextState (newState: State): void {
+    const isValidLength = (values: Array<RawParam>, index: number) => {
+      return (values[index].type === 'Balance' ? checkValueBitLength(values[index].value as BN) : true);
+    };
```
With that however, the bitlength check should be elsewhere, not specifically here. If we do it at the base, we don't need to add the same logic to each component that uses it.

this component should not care about lengths and being valid, what is passed to it should already have the required checks applied.

jacogr commented on packages/app-extrinsics/src/Extrinsic.tsx:
```diff
@@ -82,7 +88,9 @@ class Extrinsic extends React.PureComponent<Props, State> {
           isValid &&
           !isUndefined(values[index]) &&
           !isUndefined(values[index].value) &&
+          isValidLength(values, index) &&
```
The component would have already done this, i.e. it cannot be `isValid` and then have an invalid length - `isValid` for the type should indicate that the value is valid and passes all checks, including length (if applicable).

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -2,41 +2,44 @@
   min?: any,
   name?: string,
   onChange: (value: string) => void,
+  onKeyDown?: (event: SUIEvent) => void,
```
Don't pass the SUIEvent up to components using this - rather deconstruct the value in here (i.e. this component knows it uses SUI), calling components should not.

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -2,41 +2,44 @@
   name?: string,
   onChange: (value: string) => void,
+  onKeyDown?: (event: SUIEvent) => void,
+  onKeyUp?: (event: SUIEvent) => void,
```
As above per the `on*` definition.

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -90,11 +96,32 @@ export default class Input extends React.PureComponent<Props, State> {
+    const { onKeyDown } = this.props;
+
+    if (onKeyDown) {
+      onKeyDown(event);
```
Here grab the values from event (i.e. shift state, key pressed) and pass only those up.

ltfschoen replied:
I've had to pass up the `event` itself in Input.tsx:
```
    if (onKeyDown) {
      onKeyDown(event);
...
```

I can't pass up destructured properties of event with the following in Input.tsx, because I'd also have to destructure and pass up the `preventDefault` method of `event`, which I haven't been able to do
```
    if (onKeyDown) {
      onKeyDown(key, preventDefault, shiftKey, (target as HTMLInputElement).value);
...
```

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -90,11 +96,32 @@ export default class Input extends React.PureComponent<Props, State> {
+    const { onKeyUp } = this.props;
+
+    if (onKeyUp) {
+      onKeyUp(event);
```
As per the above comment, make the callers not know about SUI.

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,58 @@
+  maxLength?: number,
+  onChange: (value: string) => void,
+  onKeyDown?: (event: SUIEvent) => void,
+  onKeyUp?: (event: SUIEvent) => void,
```
SUIEvent as per previous comments.

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,58 @@
+  label?: any,
+  maxLength?: number,
+  onChange: (value: string) => void,
+  onKeyDown?: (event: SUIEvent) => void,
```
SUIEvent as per previous comments.

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+
+import Message from 'semantic-ui-react/dist/commonjs/collections/Message/index';
+import classes from './util/classes';
```
Extra blank line before this line, i.e. split external and per-package imports into seperate sections.

jacogr commented on packages/ui-app/src/Params/Param/Balance.spec.js:
```diff
@@ -0,0 +1,93 @@
+// of the ISC license. See the LICENSE file for details.
+
+import React from 'react';
+import { mount } from '../../../../../test/enzyme';
```
Import split, blank line before.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,158 @@
+          onKeyDown={this.onKeyDown}
+          onKeyUp={this.onKeyUp}
+          placeholder={t('account.balance.placeholder', {
+            defaultValue: 'Between 0 DOT and the available DOT balance of the account'
```
Remove the DOT ref. "Positive number" would convey the message just as well.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  onKeyUp = (event: any): void => {
+    if (KEYS_PRE.includes(event.keyCode)) {
```
Just pass the keyCode in from the underlying component. (Additionally, we now have an untyped `: any`, which we really should not have unless it is an exceptional circumstance.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,158 @@
+    }
+  }
+
+  onKeyDown = (event: any): void => {
```
`any`

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,158 @@
+    // only allow user balance input to contain one instance of '.' for decimals.
+    // prevent use of shift key
+    if (
+      (keydown.isDuplicateDecimalPoint(event)) ||
```
Rather not. See the comments about deconstructing the event, passing shift & keyCode. (So only the component ever wrapping SUI knows about SUI.) Standard React events are fine, not SUI-specific events.

jacogr commented on packages/ui-app/src/types.d.ts:
```diff
@@ -25,3 +25,7 @@ export type BaseContext = {
 };
+
+export type SUIEvent = {
+  value: any
```
Don't use `any` unless there is no option. Specify or use standard React events. (They are typed)

jacogr commented on packages/ui-app/src/util/chainSpec.ts:
```diff
@@ -0,0 +1,30 @@
+ * and may be entered as 3.4e38, 3.4e+38 or 340282366920938463463374607431768211455. we may
+ * show it is equivalent to 340282366920938463463374607431768211455 by calling .toString(10)
+ */
+const maxValue = (bitLength?: number): BN => new BN(2).pow(new BN(bitLength || BIT_LENGTH_128)).subn(1);
```
Split over 2 lines for readability, i.e.

```
const maxValue = (bitLength: number = BIT_LENGTH_128): BN =>
  new BN(2).pow(new BN(bitLength)).subn(1);

jacogr commented on packages/ui-app/src/util/chainSpec.ts:
```diff
@@ -0,0 +1,30 @@
+ */
+const maxValue = (bitLength?: number): BN => new BN(2).pow(new BN(bitLength || BIT_LENGTH_128)).subn(1);
+
+const maxLength = (maxValue: BN): number => maxValue.toString().length; // returns 39 for 128 bit arg
```
Readability imporvements -

```
const maxLength = (maxValue: BN): number =>
  maxValue.toString().length;
```

Additionally, we want to have one less digit, i.e. let's take 16-bit values, i.e. 2^16 = 65536. Your calculation would return 5 digits, however when I enter `70000` it is an invalid number. Err on the side of conservativeness rather. (Which is fine with 64/128/256 bit numbers)

jacogr commented on packages/ui-app/src/util/chainSpec.ts:
```diff
@@ -0,0 +1,30 @@
+
+// fixes error when enter value like 0.1111111111111111111111111111111, which causes
+// bn.js error in toU8a.js of 'byte array longer than desired length'
+const checkValueBitLength = (value: BN, bitLength?: number): boolean => {
```
Would read better as

```
const checkValueBitLength = (value: BN, bitLength?: number): boolean =>
  value.bitLength() <= (bitLength || BIT_LENGTH_128);
```

The `<=` already returns a boolean result, no need to check it.

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  const maxBN = maxValue(bitLength);
+
```
Extra empty line, remove.

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,73 @@
+        }
+      })
+    };
+  } else {
```
No need for the nested if, this is simply -

```
} else if (inputBN.isZero()) {
...
}
```

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,73 @@
+    }
+  }
+
+  const infoMessage = Number(input) <= MAX_SAFE_INTEGER ? Number.parseFloat(input).toExponential(2) : undefined;
```
Readability -

```
const infoMessage = Number(input) <= MAX_SAFE_INTEGER
  ? Number.parseFloat(input).toExponential(2)
  : undefined;
```

However, we are still using Number when dealing with BN. The whole idea is that we only use `BN` checks, never anything from `Number`.

jacogr commented on packages/ui-app/src/util/keyboard.ts:
```diff
@@ -0,0 +1,32 @@
+// obtain current cursor index position of an input field
+const cursorIndexInputField = (event: any) => event.target.value.slice(0, event.target.selectionStart).length;
+
+const isNotDigit = (event: any): boolean => (event.keyCode < 48 || event.keyCode > 57);
```
For the whole file, split the definition of the function from the actual body. (See readability improvement comments earlier)

jacogr commented on packages/ui-app/src/util/types.d.ts:
```diff
@@ -12,3 +14,10 @@ export type DropdownOption = {
+
+export type IsValidWithMessage = {
+  isValid: boolean,
+  errorMessage?: TranslationFunction,
```
Only strings. Not sure why we are padding the definition of `t` in here.

ltfschoen commented:
Progress Update:

Remaining review comments:

* [ ] - Don't pass the SUIEvent up to components using this - rather deconstruct the value in here (i.e. this component knows it uses SUI), calling components should not.
* [ ] -  As above per the on* definition.
* [ ] -  Here grab the values from event (i.e. shift state, key pressed) and pass only those up.
* [ ] -  As per the above comment, make the callers not know about SUI.
* [ ] - SUIEvent as per previous comments.
* [ ] - SUIEvent as per previous comments.
* [ ] - Just pass the keyCode in from the underlying component. (Additionally, we now have an untyped : any, which we really should not have unless it is an exceptional circumstance.
* [ ] - any
* [ ] - Rather not. See the comments about deconstructing the event, passing shift & keyCode. (So only the component ever wrapping SUI knows about SUI.) Standard React events are fine, not SUI-specific events.
* [ ] - Don't use any unless there is no option. Specify or use standard React events. (They are typed)
* [ ] - Readability imporvements -
const maxLength = (maxValue: BN): number =>
  maxValue.toString().length;
Additionally, we want to have one less digit, i.e. let's take 16-bit values, i.e. 2^16 = 65536. Your calculation would return 5 digits, however when I enter 70000 it is an invalid number. Err on the side of conservativeness rather. (Which is fine with 64/128/256 bit numbers)

ltfschoen commented:
@jacogr I've pushed commits that address all your review comments. Note that I've replied directly to your comment "Here grab the values from event (i.e. shift state, key pressed) and pass only those up.".

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,58 @@
+  withLabel?: boolean
+};
+
+class InputNumber extends React.PureComponent<Props> {
```
Ok, looking this as an "InputNumber". It is just a text input. As a number input, I would expect it to only take numbers. So the logic we have in Balance that does actual numbers should be in here. The only addition Balance should do is to convert to a Param.

1. This component should do number input checks
2. This component should convert to a number (BN in our case)
3. Users of this can further enhance the values

So it is just a move of logic to the base component, i.e. this one.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,158 @@
+      }
+
+      let valueBN: BN;
+      valueBN = new BN(value || '0');
```
This type of thing in `InputNumber` as per comment above.

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,158 @@
+    if (keydown.isNonNumeric(event.key)) {
+      this.setState({
+        error: t('balance.error.format', {
+          defaultValue: 'Balance to transfer in DOTs must be a positive number'
```
Balance does not have anything to do with "transfer" - transfer is just one extrrinsic, a number of others also take Balances as inputs.

ltfschoen commented:
@jacogr I'm on a plane that's about to take off, but I've pushed a commit that I think addresses what you mean, could you take a look and let me know if i'm on the right track?

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,159 @@
+    super(props);
+
+    this.state = {
+      error: '',
```
(nitpick) Since these are marked as optional, we actually don't need to assign values to them here.

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,159 @@
+    if (keydown.isNonNumeric(event.key)) {
+      this.setState({
+        error: t('balance.error.format', {
+          defaultValue: 'Balance in DOTs must be a positive number'
```
Can't we just reject the keys outright? As a normal `type='number'` HTML input behaves. (Not a major issue, but not allowing is always better than allowing and then saying "you are doing it wrong")

jacogr replied:
EDIT: This will make the `isValid` pass-up not needed as well since we will always have a valid value here.

jacogr replied:
Additionally, This is a number Input, not a Balance input - so a number can be anything, not just a DOT. (Also remove the DOT references everywhere, for substrate chains they have their own tokens)

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,159 @@
+  isError?: boolean,
+  label?: any,
+  maxLength?: number,
+  onChange?: RawParam$OnChange | undefined,
```
This should just be a `(value: BN) => void` callback. The Param stuff is only for Params. (e.g. see other Inputs, it only returns the value)

ltfschoen replied:
I've had to change it to `onChange?: (value: { isValid: boolean, value: BN }) => void,` instead of `onChange?: (value: RawParam$OnChange$Value) => void,` (that way we're not using Param types), and because the onChange that is passed as a prop to InputNumber from the Balance component comes from BareProps and is of type `RawParam$OnChange | undefined`, and the `onChange` method that we call on InputNumber returns `{ isValid, value: valueBN }`

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,159 @@
+      }
+
+      onChange({
+        isValid,
```
As per earlier comments on the actual onChange type - we should check the value, if valid, callback with _only_ the value, nothing more required.

jacogr replied:
This is still outstanding. `onChange` should be - `onChange: (value: BN) => void`, not extra `inValid` is should be what it says on the tin, i.e. a number input.

jacogr commented on packages/ui-app/src/constants.ts:
```diff
@@ -0,0 +1,42 @@
+
+const KEYS_PRE: Array<any> = [KEYS.ALT, KEYS.CMD, KEYS.CTRL];
+
+const MAX_SAFE_INTEGER: number = 9007199254740991; // 2**53-1
```
Rathe `const MAX_SAFE_INTEGER: number = 2 ** 53-1;` instead of working it out and commenting on it. Additionally, not convinced at all that we need to re-define what JS already gives us - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

jacogr commented on packages/ui-app/src/util/chainSpec.ts:
```diff
@@ -0,0 +1,39 @@
+ */
+const maxLength = (maxValue: BN): number => {
+  const conservativenessFactor = 1;
+  return maxValue.toString().length - conservativenessFactor; // returns 38 for 128 bit arg
```
(nitpick) blank line before this line.

jacogr commented on packages/ui-app/src/util/isValidBalance.ts:
```diff
@@ -0,0 +1,90 @@
+    };
+  }
+
+  const maxSafeIntegerBN = new BN(String(MAX_SAFE_INTEGER));
```
Just simply `new BN(Number.MAX_SAFE_INTEGER)` - no coercions, no re-definition of built-in values.

jacogr commented on packages/ui-app/src/util/keyboard/isNonNumeric.ts:
```diff
@@ -0,0 +1,8 @@
+
+export default function isNonNumeric (key: string): boolean {
+  const digits = new Array(9).join().split(',').map((item, index) => String(++index));
+  return !digits.includes(key);
```
(nitpick) blank line before

jacogr commented on packages/ui-app/src/util/keyboard/isDuplicateDecimalPoint.ts:
```diff
@@ -0,0 +1,14 @@
+  const inputValue: string = value;
+  const didPressDecimalPoint: boolean = key === KEYS.DECIMAL_POINT;
+  const foundExistingDecimalPoint: boolean = inputValue.match(regexDecimalPoint) ? true : false;
+  return didPressDecimalPoint && foundExistingDecimalPoint;
```
(nitpick) blank line before

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,52 @@
+      </Bare>
+    );
+  }
+}
```
Great. Now, this component should make it a valid RawParam, i.e. take the BN and add the isValid stuff as required. (Which in most cases should be just always `true`, i.e. what we receive from the InputNumber should be a valid number based on the logic there)

jacogr commented on packages/ui-app/src/util/isValidBalance.spec.js:
```diff
@@ -0,0 +1,23 @@
+
+import { BIT_LENGTH_128 } from '../constants';
+import isValidBalance from './isValidBalance';
+import translate from '../translate';
```
(nitpick) move translate one up, so external to folder first, then internal to folder.

ltfschoen commented:
@jacogr I've tried to address all the review comments. I've left it passing translate down as props from Balance to InputNumber because otherwise I had difficulty getting the unit tests to remain working (both when I wrapped InputNumber with translate, and when I tried creating an InputNumber.test.js and modified version of the same unit tests on that component instead)

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,151 @@
+  }
+}
+
+export default InputNumber;
```
I see the testing comment, however when we specify `t` the component needs to be wrapped in a translate. We cannot have components using this pass through `t`. So it _needs_ to be wrapped. (The testing setup is incorrect if it is not working, i.e. i18next is not being initialised)

ltfschoen commented:
@jacogr I've overhauled the code and tried to make it more organised.

I've addressed both your review comments (i.e. only passing the value to `onChange` without isValid by ensuring I make the value valid in the InputNumber component itself, and I'm no longer passing translate down from Balance, I'm just injecting the observable directly into InputNumber and moved the tests into InputNumber and got them working).
I've written a detailed list of other changes in this commit, where most of the work was done https://github.com/polkadot-js/apps/pull/221/commits/ba7619b2a0109dc4c89dab1c0606fb41a7c0cbc6

jacogr commented:
Ok, As mentioned above. Can be please trim this down and get rid of unneeded stuff that makes it flashy without real benefit -

1. I'm not sure why this is even here. This is the first thing that my non-techy family and friends will ask about. This adds no value at all to the users.

<img width="276" alt="polkadot apps portal 2018-09-19 14-08-45" src="https://user-images.githubusercontent.com/1424473/45752550-1efd0180-bc16-11e8-87a7-da3f62d93f25.png">

2. Not sure why this is here - in this case you can still submit, the warning just looks out of place.

<img width="276" alt="polkadot apps portal 2018-09-19 14-09-10" src="https://user-images.githubusercontent.com/1424473/45752530-13a9d600-bc16-11e8-8998-19c8329619da.png">

3. Since actual transfers will be done in the transfer app, we need to get it in there -

<img width="218" alt="polkadot apps portal 2018-09-19 14-09-47" src="https://user-images.githubusercontent.com/1424473/45752609-5ec3e900-bc16-11e8-9170-298d11da8069.png">

4. I'm struggling to see where we are actually now doing errors in the Balance input - it filters everything non-valid.

5. Move all the different files that are only used in the number input into functions inside that input. Then we also get rid of calling external functions (like isNumberValid) with t as a param. Keep it small, keep it together.

ltfschoen commented:
Progress update:

* ASSUMPTION: InputNumber may be used in future for useful error, warning, and info messages, so leave the boilerplate props along with Notification.

* ASSUMPTION: Displaying an info message in scientific notation i.e. 3e+7
causes more confusion than benefit since most users are non-techy. In future this when we have Issue #86 Settings App, it could be added as an optional extra that could be toggled by the user along with Issue #224 E-Notation
	* [x] - Remove display of scientific notation equivalent
	* [x] - Remove warning 'number.warn.zero' when user enters 0 value
	* [x] - Remove info notification 'number.info.scientific' when user enters valid value
	* [x] - Remove `toScientificNotation` function, which is incorrectly passed the TranslationFunction `t` anyway.
	* [x] - Remove `IsValidWithMessage` type and replace return value with just a boolean of whether its valid or not
	* [x] - Move `isValidNumber` into React componet, and do not pass the TranslationFunctio to it!
	* [x] - Move `isValidKey` into React componet, and do not pass the TranslationFunctio to it!
	* [x] - Move `isNonDecimal` into React componet, and do not pass the TranslationFunctio to it!
	* [X] - Remove `isNonInteger` and isNonInteger.ts and isNonInteger.spec.js since no longer using
	* [X] - Remove `error`, `warn`, `info` from InputNumber state, since we're filtering out everything not valid. Instead add them as optional props (so Balance or another component can display notifications if required)
	* [ ] - Move tests for isValidKey into React component, and remove isValidKey.spec.js
	* [ ] - Move tests for isValidNumber into React component, and remove isValidNumber.spec.js
	* [ ] - Move tests for isNonDecimal into React component, and remove isNonDecimal.spec.js
	* [ ] - Update InputNumber.spec.js to remove useless tests that check that the Notification component isn't shown
	* [ ] - Integrate into the Transfer App `send a value of` where actual transfers will be done (see @jacogr comment and screenshot)
	* [ ] - Move all the different files that are only used in the number input into functions inside that input. Then we also get rid of calling external functions (like isNumberValid) with t as a param. Keep it small, keep it together.
	* [ ] - Make methods `private` as relevat
	* [ ] - Figure out what action to take yet based on @jacogr's comment: "I'm struggling to see where we are actually now doing errors in the Balance input - it filters everything non-valid.". 
	Maybe the Balance isValid needs an isBalanceValid function that checks if the user has sufficient balance and passes error as props to InputNumber?
	* [ ] - Figure out what action to take yet based on @jacogr's comment:  "Seems SUI docs are currently broken. But in the cases where you absolutely need to point to an input with an attachment, their Label does exactly that and allows attachment with a pointer to an Input. So at most Notification should just warp that. And it really should only be used is completely exceptionally circumstances. (And I really couldn't see any such circumstances from playing with the actual inputs. So at this point seems like a good, although unused in practice idea.)"

jacogr commented:
Second assumption is wrong - it is about allowing the "input of scientific notation number", not the display of them on an otherwise clean input. Basically, the user (in the future), will be able to input either

`1000000` or `1e6` and both would be accepted.

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,188 @@
+  defaultValue?: string,
+  isError?: boolean,
+  label?: any,
+  maxLength?: number,
```
I believe we need to pass in bitLength here instead of maxLength. Additionally, put that at 32-bits default (i.e. which is just "normal" numbers, for Balance and friends we will pass through the correct length as required.

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,188 @@
+        <Input
+          {...this.props}
+          defaultValue={defaultValue || '0'}
+          maxLength={maxLength || this.maxLength()}
```
Here maxLength is fine, however we use the bitLength as passed in to calculate to pass to Input. So InpoutNumber works with bitLength, the lower-level can work with character lengths.

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,188 @@
+    new BN(2).pow(new BN(bitLength || BIT_LENGTH_128)).subn(1)
+
+  private maxLength = (): number => {
+    const conservativenessFactor = 1;
```
You can put this one file-level since it never changes. (like you have the default bitlength)

jacogr commented on packages/ui-app/src/Params/Param/Balance.tsx:
```diff
@@ -0,0 +1,50 @@
+    return (
+      <Bare {...this.props}>
+        <InputNumber
+          {...this.props}
```
Not sure why we are passing the same props to both Bare and InputNumber - rather be explicit here as to what goes where.

jacogr commented on packages/ui-app/src/InputNumber.tsx:
```diff
@@ -0,0 +1,191 @@
+  previousValue: string
+};
+
+const BIT_LENGTH_32 = 32; // normal numbers
```
Any reason we can't just use 32, 128 where we pass it in?

Elsewhere in the API (encoding/decoding), have actually defined this type of thing in the following manner:

```
type BitLength = 32 | 64 | 128 | 256;

// default here
function foo (bitLength?: BitLength = 128) {
 ...
}

foo(32); // works
foo(64); // works
foo(121); // does not work, TS checks shows an error
```

ltfschoen replied:
@jacogr I've pushed a commit with these changes.

I also created a TS `enum` in enum.ts (since enums don't go in .d.ts files) and used it as shown below, but it wouldn't recognise it in the unit tests

```
// enum.ts
export declare const enum BitLengthOption {
  CHAIN_SPEC = 128,
  NORMAL_NUMBERS = 32
}

// InputNumber.tsx
const DEFAULT_BITLENGTH: BitLength = BitLengthOption.NORMAL_NUMBERS;

// Transfer.tsx / Balance.tsx
const DEFAULT_BITLENGTH: BitLength = BitLengthOption.CHAIN_SPEC;
```

So I reverted to just using key/value pairs, which worked and passes the unit tests. But I wasn't sure where to put it (i.e. if I put it in Balance.tsx, then it would be strange to import it and use BitLengthOption in Transfer.tsx where we are importing InputNumber directly...), so I've just left it in enum.ts for the moment.

```
// enum.ts
export const BitLengthOption = {
  CHAIN_SPEC: 128 as BitLength,
  NORMAL_NUMBERS: 32 as BitLength
};
```

I like doing this because I find it more readable (and avoids me leaving a comments like `// 32 is normal number, 128 is chain spec` above the type definition of BitLength for future reference, but I'm also happy to remove it

ltfschoen commented:
@jacogr I've pushed a commit with these changes.

I also created a TypeScript `enum` in enum.ts (since enums don't go in .d.ts files, but I wasn't sure where would be an appropriate place for it). I wasn't sure where to put it because if I put it in Balance.tsx, then it would be strange to import it and use BitLengthOption in Transfer.tsx where we are importing InputNumber directly (instead of Balance), so I've just left it in enum.ts for the moment until I get your feedback, and whether to use the enum at all

I like using the `enum` here because I find it more readable (and avoids me leaving a comments like `// 32 is normal number, 128 is chain spec` above the type definition of BitLength for future reference, or writing long variable names, but I'm also happy to remove it

ltfschoen commented:
@jacogr I've pushed a commit addressing your review comments to use `BitLength` type.

I also created a TypeScript `enum` named `BitLengthOptions` in constants.ts so we can give meaningful names for their usage. I put it in constants.ts because I wasn't sure where to put it because if I put it in Balance.tsx, then it would be strange to import it and use BitLengthOption in Transfer.tsx where we are importing InputNumber directly (instead of Balance), so I've just left it in constants.ts for the moment until I get your feedback, and whether to use the enum at all

I like using the `enum` here because it augments the use of constants to make it more readable (and avoids me leaving a comments like `// 32 is normal number, 128 is chain spec` above the type definition of BitLength for future reference, or writing long variable names, but I'm also happy to remove it

ltfschoen commented:
@jacogr I've pushed an additional commit that I believe solves the problem of preventing the user from entering a value that's preceded with 0, which is SOLUTION OPTION 3 below (where I've also included other options I considered. 

* RESTRICTED user actions
  * user should not be able to enter a value that displays multiple 0's (i.e. 00), even though a display value of '00' would be submitted as '0', on the basis that allowing it looks unprofessional
    * currently we restrict the user from entering the key 0 when the existing value is 0
  * user should not be allowed to enter or paste a value that is preceeded by 0 (i.e. '01', or '01234'), because this display value would actually be submitted as '1' or '1234' (this could be dangerous because a user may think 01234 would be interpreted as just 0, and actually send/transfer more value than they anticipated!)
    * OPTION 1 - revert to the previous value if the first digit of the new value in the input field is '0' (after they enter a key on the RHS of the previous value of '0', or they paste a value after a previous value of '0', or they enter a key or paste a value that replaces the whole value with a value that is preceded with '0')
      * Cons - when implemented in onKeyDown it produces an ugly UI, it allows the user to enter the wrong values and displays them, and only reverts the changes after they keyup
      * Verdict - IGNORE
    * OPTION 2 - prevent the user from pasting a value if their cursor position is on the RHS of a '0' that is the first character of the input value
      * Cons - confuses the user and they may think they can't paste at all, not realising they can paste on the LHS
      * Verdict - IGNORE
    * **OPTION 3** - allow the user to enter a value or paste, but strip all the preceding '0's from the new value, which can be achieved in keydown. and as the user may have entered or pasted a value of '0' so it became '00' we need to ensure that we're not stripping the value away entirely, so only strip if the new value isn't still a zero value
      * **SOLUTION**:
        * **implemented in onKeyUp**

```
/* if new value equates to '0' in BN when but it's length is >=1 (i.e. '012', '00', etc)
 * then replace the input value with just '0'.
 * otherwise remove the preceding zeros from the new value (i.e. '0123' -> '123')
 * note: edge case glitch occurs if existing value is '0' and you 'hold down' and keep
 * pasting a value of '00' after it, then sometimes when you let go the
 * remaining value shown as '000' or '00000' in the UI, but it's still ok because
 * the actual BN if the user submitted would still be '0', and if they then press any key
 * the UI input value resets to '0'
 */
 (event.target as HTMLInputElement).value = (isNewValueZero && !!newValue.length) ? '0' : newValue.replace(/^0+/, '');
```
        * **implemented in onKeyDown**

This solution is augmented by in onKeyDown which prevents the user from entering the '0' key if the digit in the first index position of the input field is already '0'

```
// prevents entry of zero if initial digit is zero
const isDuplicateZero = previousValue[0] === '0' && event.key === KEYS.ZERO;

if (isDuplicateZero) {
  return false;
}
```


* PERMITTED user actions
  * user should be able to change the value from '0' to '', because even though it displays '', if they press submit then '0' will be submitted (so it's not necesssary to propagate isValid up from InputNumber via onChange to Balance or Transfer. assumes most users know that if a submission button is disabled it means an input is invalid, so to avoid confusion they just need to follow the instruction of entering a 'Positive Number'
    * otherwise we'd have to either:
      * prevent them deleting the '0', which would be horrible as they'd have to move their cursor to index 0 before entering any value
        * unless we automatically removed the '0' value and replaced it with the value they entered (which may confuse some users)

ltfschoen commented on packages/ui-app/src/types.d.ts:
```diff
@@ -25,3 +25,5 @@ export type BaseContext = {
   }
 };
+
+export type BitLength = 32 | 64 | 128 | 256;
```
@jacogr Should this be `8 | 16 | 32 | 64 | 128 | 256` so its the same as https://github.com/polkadot-js/api/commit/60d3d6396a8e48626c74376e5cba0bdd41f17110#diff-0dce4593d48039d48c497cf070a879cfR16

jacogr replied:
I think (and we are going through this in the encoding atm), that all numbers are (currently) 32 (most indexes, i.e. Nonce, PropIndex, ReferendumIndex, ...) or 64 (BlockNumber, ...) or 128 (Balance). We actually don't have any values shuttled around that don't fit here.

For future-proofing, the 8 | 16 | ... is probably better. (Although it can be added as the need arrises.)

jacogr replied:
TL;DR - We can adpat either way, but happy as it is, and won't complain if it is changed either.
